File: Go/bitwise_xor.go
/*
╔════════════════════════════════════════════════════════╗
║              Bitwise XOR (Single Number)              ║
╠════════════════════════════════════════════════════════╣
║ Description:                                           ║
║ Use XOR to find the single number in an array where   ║
║ all other numbers appear twice. XOR cancels out       ║
║ duplicates: a^a=0, a^0=a.                            ║
╠════════════════════════════════════════════════════════╣
║ Flow Diagram (ASCII):                                  ║
║   nums = [4,1,2,1,2]                                  ║
║                                                      ║
║   result = 0                                          ║
║   0 ^ 4 = 4                                           ║
║   4 ^ 1 = 5                                           ║
║   5 ^ 2 = 7                                           ║
║   7 ^ 1 = 6                                           ║
║   6 ^ 2 = 4 → single number found                     ║
╚════════════════════════════════════════════════════════╝
*/

package main

import "fmt"

func singleNumber(nums []int) int {
    result := 0
    for _, num := range nums {
        result ^= num
    }
    return result
}

func main() {
    nums := []int{4, 1, 2, 1, 2}
    single := singleNumber(nums)
    fmt.Printf("Single number: %d\n", single)
} 

File: Python/bitwise_xor.py
"""
╔════════════════════════════════════════════════════════╗
║               Bitwise XOR Pattern (Single Number)     ║
╠════════════════════════════════════════════════════════╣
║ Description:                                           ║
║ XOR all numbers; since x⊕x=0 and x⊕0=x, pairs cancel ║
║ leaving the unique number.                            ║
╠════════════════════════════════════════════════════════╣
║ Flow Diagram (ASCII):                                  ║
║ arr=[4,1,2,1,2]                                       ║
║  res=0⊕4=4;                                           ║
║  res=4⊕1=5;                                           ║
║  res=5⊕2=7;                                           ║
║  res=7⊕1=6;                                           ║
║  res=6⊕2=4;                                           ║
║  → result=4                                           ║
╚════════════════════════════════════════════════════════╝
"""

def single_number(nums):
    result = 0
    for x in nums:
        result ^= x
    return result

def main():
    arr = [4,1,2,1,2]
    print(f"Single number = {single_number(arr)}")

if __name__ == "__main__":
    main() 